src/errors/error_kind.rs

```diff
@@ -0,0 +1,319 @@
+
+    // Serialization
+    #[error("JSON: {:?}", .0)]
+    SerializeJSON(serde_json::Error),
+    #[error("x-form-urlencoded: {:?}", .0)]
+    SerializeForm(#[from] serde_urlencoded::ser::Error),
+}
+
+impl From<BodyError> for Error {
+    fn from(other: BodyError) -> Self {
+        Error::Body(other)
+    }
+}
+
+#[derive(Debug, Diagnostic, ThisError)]
+#[non_exhaustive]
+pub enum HeaderError {
+    // Parse
+    #[error("Header value specificity was invalid")]
+    SpecificityInvalid,
+    /// Header name was invalid
+    #[error("Header name was invalid: {}", .0)]
+    NameInvalid(&'static str),
+    /// Header value was invalid
+    #[error("Header value was invalid: {}", .0)]
+    ValueInvalid(&'static str),
+
+    // Parse specific headers
+    #[error("Date header was invalid: {}", .0)]
+    DateInvalid(DateError),
+    #[error("No suitable Transfer-Encoding found during negotiation")]
+    TransferEncodingUnnegotiable,
+    #[error("Transfer-Encoding header encoding was invalid: {}", .0)]
+    TransferEncodingInvalidEncoding(EncodingError),
+    #[error("Trace-Context header was invalid: {}", .0)]
+    TraceContextInvalid(&'static str),
+    #[error("Server-Timing header was invalid: {}", .0)]
+    ServerTimingInvalid(&'static str),
+    #[error("Server-Timing header metric was invalid: {}", .0)]
+    ServerTimingInvalidMetric(&'static str),
+    #[error("Timing-Allow-Origin header was invalid: {:?}", .0)]
+    TimingAllowOriginInvalidUrl(url::ParseError),
+    #[error("Forwarded header was invalid: {}", .0)]
+    ForwardedInvalid(&'static str),
+    #[error("Sourcemap header url was invalid: {:?}", .0)]
+    SourceMapInvalidUrl(url::ParseError),
+    #[error("Sourcemap header base url was invalid: {:?}", .0)]
+    SourceMapInvalidBaseUrl(Box<dyn std::fmt::Debug + Send + Sync + 'static>),
+    #[error("Referer header url was invalid: {:?}", .0)]
+    RefererInvalidUrl(url::ParseError),
+    #[error("Referer header base url was invalid: {:?}", .0)]
+    RefererInvalidBaseUrl(Box<dyn std::fmt::Debug + Send + Sync + 'static>),
+    #[error("Content-Type header MediaType (MIME) was invalid: {}", .0)]
+    ContentTypeInvalidMediaType(MediaTypeError),
+    #[error("Content-Length header was invalid: length out of bounds (unsized 64 bit integer)")]
+    ContentLengthInvalid,
+    #[error("Accept header was invalid: {}", .0)]
+    AcceptInvalidMediaType(MediaTypeError),
+    #[error("No suitable Content-Type header MediaType found during Accept negotiation")]
+    AcceptUnnegotiable,
+    #[error("Accept-Encoding header was invalid: {}", .0)]
+    AcceptEncodingInvalidEncoding(EncodingError),
+    #[error(
+        "No suitable Content-Encoding header Encoding found during Accept-Encoding negotiation"
+    )]
+    AcceptEncodingUnnegotiable,
+    #[error("ETag header was invalid")]
+    ETagInvalid,
+    #[error("Age header was invalid: length out of bounds (unsized 64 bit integer)")]
+    AgeInvalid,
```

Comment:
In general, do we want to have one of these variants for every header, or do we want to have one variant that has a header name and a message?

Reply:
Another question would be naming... how do we feel about `<header name>Invalid<optional detailed invalid area>`?